### DISCUSSION TITLE
🌟 Universal Functions and new plot types 🌟

### DISCUSSION THREAD

**Discussion by dbuezas** [dbuezas/lovelace-plotly-graph-card#204] [Announcements]
# The biggest and most complex release of the year!
## ~~Open for beta testing!~~

(v3.0.0 released)

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/777196/215200293-d2655190-1ae3-458f-a42a-cb85383b67db.png">

* YAMLs of these examples [here](https://gist.github.com/dbuezas/92210c8fcaf8d86d6c2d358e9d83b52d)
* WIP readme [here](https://github.com/dbuezas/lovelace-plotly-graph-card/blob/experiment/universal-fn/readme.md)

# What's Changed

## Improvements

* Universal functions allow you to communicate across the yaml, combine datapoints from multiple entities into one, add the state of a sensor to the name of the trace, etc. 
* `show_value` now can print out of the plotting area
* Semi-persistant cache: resetting the plot doesn't delete data anymore
* Efficiency: offset entities share the same cache with non-offset ones.
* Added almost all plotly plot types, including `histograms`, `3dscatter`, `surface`, `heatmap`, `piechart`, ... Find inspiration [here](https://plotly.com/javascript/) (maps not included)
* Added a delta filter (more to come!)
* New `raw_plotly_config` yaml option to disable all in-built defaults (useful for non-standard plots)
* Home Assistant themes' color variables available inside Universal Functions (`$fn`)
* Respect fixed range axes when pinch zooming
* Improved pinch to zoom algorithm
* Universal `$fn` functions can access other values from the yaml, even if those are also `$fn` computed.
* UI Informs of out of order `$fn` dependencies.
* `$fn` execution is strictly defined (top to bottom, shallow to deep, aka depth first)
* Errors in the yaml or functions don't interrupt plotting but are shown in the UI with intuitive phrasing. 
* ...

## Breaking Changes

The UI will let you know which plots are affected, how, and what to do.

* `no_theme`:  renamed to ha_theme (inverted logic) 
* `no_default_layout`:  replaced with more general raw_plotly_config 
* `offset`:  renamed to time_offset  to avoid conflicts with [bar-offset](https://plotly.com/javascript/reference/bar/#bar-offset)
* `lambda`:  removed, use filters instead
* `significant_changes_only`:  removed, it is now always set to false
* `minimal_response`:  removed, if you need attributes use the 'attribute' parameter instead.
* `show_value.right_margin`: removed, the values are shown out of the plotting area. use time_offsets to make space for them

## Beware: beta version, APIs may change

Here is [how to enable beta releases](https://hacs.xyz/docs/faq/beta/) in HACS

**Full Changelog**: https://github.com/dbuezas/lovelace-plotly-graph-card/compare/v2.1.0...v3.0.0-beta

<hr /><em>This discussion was created from the release <a href='https://github.com/dbuezas/lovelace-plotly-graph-card/releases/tag/v3.0.0-beta'>Universal Functions (`$fn`) and new plot types</a>.</em>

**Comment by MathieuIci** [comment]
Awesome release, thank you!

I no longer need to use the lovelace-card-templater 🤓

I discovered this card a few days ago and have been having lots of fun, particularly with the "fn" filter:

![image](https://user-images.githubusercontent.com/57320163/215238511-013fa246-b742-4911-ba8b-985913a0032a.png)

I have two questions:

1.  With this new beta release, I had to add `ha_theme: true` to my card config for the Home Assistant dark theme to be applied.  Is this expected?

2. I've gone trough the documentation many times and searched online for an answer, but couldn't figure out if there are any way to define "tap" actions on title/annotations/icons (/anything) to have them perform more-info, service call, etc. ?  This would bring so many more possibilities, it's something I miss from the ApexCharts card.

Edit: Hmm, I think something like [this feature enhancement request](https://github.com/dbuezas/lovelace-plotly-graph-card/issues/72) is what I am looking for.

Thanks again!

**Reply by dbuezas** [reply]
Happy to hear!

1. Woops. Small last mile mistake. Fixed `ha_theme` in v3.0.0-beta2.
2. Not possible, drop a feature request issue! :)  
3. That's a pretty plot you have there! Please consider a small post with yaml & screenshot in [Show&Tell](https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/categories/show-and-tell) :)

**Reply by MathieuIci** [reply]
1. Thanks, I updated to 3.0.0-beta2 and it now works without having to use `ha_theme` :slightly_smiling_face: 
2. This feature would be very close to existing feature request [#72](https://github.com/dbuezas/lovelace-plotly-graph-card/issues/72 ), so I'll add comments to that thread.
3. Thanks :blush:  It's a work in progress but I sure will when it's done!

I have been trying to add a button to the ModeBar by using [this information](https://plotly.com/javascript/configuration-options/#add-buttons-to-modebar) and adding this yaml to the card's configuration:
```
config:
  modeBarButtonsToAdd:
    - name: testbutton
      icon: Plotly.Icons.pencil
      click: function(gd) {alert('Button clicked')}
```
But when I do I get this error in the console's logs: `Uncaught (in promise) Error: must provide button 'click' function in button config`.  Am I doing it wrong or is the ability to define extra buttons not supported by the card?  If it were possible, could we then access the hass object and perform various HA actions when those extra buttons are clicked? (I can smell another feature request coming :wink:)

**Reply by dbuezas** [reply]
Oh, you can't directly put javascript in yaml. But you are on the right track!
Try `click: $fn () => function(gd) {alert('Button clicked')}`
The card will evaluate the outer function while parsing and you'll have a real function as value.
The issue is that there isn't much you can do inside the function, but I could add some hooks to control the plot. What do you plan to do with that? 

And regarding #72, yes it similar indeed! I even made a branch that does it when you click on the legend: https://github.com/dbuezas/lovelace-plotly-graph-card/tree/feat/entity-info-on-legend-click/src

I abandoned it at some point because it collided with toggling trace visibility, but I can add it back after finishing this branch if you need it.

**Reply by MathieuIci** [reply]
Yes, that works!  Thanks!

Just for fun and testing, I used this:

```
config:
  modeBarButtonsToAdd:
    - name: turnLightOn
      click: >-
        $fn ({hass}) => {thishass = hass; return function(gd) {
        thishass.callService('light', 'turn_on', {entity_id:
        'light.portique_lumiere'}   ) }}
```

And now the graph has a button that can turn a light on :laughing: 

Ultimate goal is to use the graph card to control heating, so I'm looking at adding controls such as "turn heat on/off" and "increase/decrease setpoint".  These buttons might not be the best method, but I'm exploring all the options.

**Reply by dbuezas** [reply]
Oh cool! I didn't think about the hass object!
Why do you have an extra "thishass" variable? "hass" is in scope of the deeper function so you shouldn't need that intermediate step

**Reply by MathieuIci** [reply]
lol, you're right!  In my fiddling it wasn't working at first so I wrongly assumed that hass wasn't in the deeper function's scope.

So, simply this does the job:
```
config:
  modeBarButtonsToAdd:
    - name: toggleLight
      click: >-
        $fn ({hass}) => function(gd) {hass.callService('light', 'toggle', {entity_id: 'light.portique_lumiere'})}
```

:partying_face: :partying_face:

**Reply by dbuezas** [reply]
Maybe I just add a click handler to legends, users can disable the toggle visibility and write a fn like the one you have. Wdyt?

**Reply by MathieuIci** [reply]
That's a great idea!

Can a function using the hass object be used to show the "more-info" card of an entity?

If the function [returns false](https://plotly.com/javascript/plotlyjs-events/#legend-click-events), it would disable toggling the visibility so that part is easy.

**Reply by dbuezas** [reply]
Oh, cool, I didn't know that. It fits perfectly then 🎉 

To trigger the more-info card, you can fire an event:
```js
fireEvent(this, "hass-more-info", {
   entityId: "sensor.blah",
});
```
and here https://github.com/custom-cards/custom-card-helpers/blob/52d1e7708be51c8262a6b94da9bc36dddd9891bc/src/fire-event.ts#L57

**Reply by dbuezas** [reply]
This works fine
```js
event = new Event( "hass-more-info")
event.detail =  { entityId: 'sensor.garden_temperature' };
document.querySelector('home-assistant').dispatchEvent(event);
```

**Reply by MathieuIci** [reply]
Awesome!  Thanks you!

Having fun...

Creating subplots for each of my temperature sensors, I'm trying to launch the more-info dialogue when clicking the **yaxis** title of each subplot.  (I prefer clicking the axis title than the legend, as when using multiple plots the legends are grouped instead of showing next to each plot)

I can run this code manually in my browser console and it works!  Clicking the title of the axis opens up the more-info dialogue :smiley: 

```
ytitleElement = document.querySelector('home-assistant').shadowRoot.querySelector('home-assistant-main').shadowRoot.querySelector('app-drawer-layout').querySelector('ha-panel-lovelace').shadowRoot.querySelector('hui-root').shadowRoot.querySelector('ha-app-layout').querySelector('hui-masonry-view').shadowRoot.querySelector('plotly-graph').shadowRoot.querySelector('svg g.g-ytitle');
ytitleElement.setAttribute("pointer-events", "all");
ytitleElement.onmouseup = function(e) {console.log('Displaying more-info'); event = new Event( "hass-more-info"); event.detail =  { entityId: 'climate.rez_de_chaussee_chauffage_thermostat' }; document.querySelector('home-assistant').dispatchEvent(event);}
```

["more-info" popup when clicking Y axis title](https://user-images.githubusercontent.com/57320163/215349250-b2d7a075-0d4e-46ef-88da-f27b9e3c78c3.webm)


But, when I put the code inside an entity's definition:

```
entities:
  - entity: climate.rez_de_chaussee_chauffage_thermostat::hvac_action
    fntest: |-
      $fn ({hass}) => {
        ytitleElement = document.querySelector('home-assistant').shadowRoot.querySelector('home-assistant-main').shadowRoot.querySelector('app-drawer-layout').querySelector('ha-panel-lovelace').shadowRoot.querySelector('hui-root').shadowRoot.querySelector('ha-app-layout').querySelector('hui-masonry-view').shadowRoot.querySelector('plotly-graph').shadowRoot.querySelector('svg g.g-ytitle');
        ytitleElement.setAttribute("pointer-events", "all");
        ytitleElement.onmouseup = function(e){console.log('Displaying more-info'); event = new Event( "hass-more-info"); event.detail =  { entityId: 'climate.rez_de_chaussee_chauffage_thermostat' }; document.querySelector('home-assistant').dispatchEvent(event);}
      }
  ...
```

It does not work.  The attribute and event are added to the element, but the parent `<g class="infolayer">` SVG element gets recreated and the changes are lost :unamused:.  These lines of code obviously belong somewhere else, for example where it would get re-executed each time the graph is redrawn, but I wouldn't know how to do that.

Do you have a suggestion?

Getting so close!  :blush:

PS: Is there an alternative to using the crazy-long element selection command?  Does HA provide variables pointing to the shadow-root of the `plotly-graph` shadow-root?

**Reply by MathieuIci** [reply]
Oooh, using the `'svg.main-svg g.xy:first-child'` selector allows us to attach the event to the whole first graph itself!  (using `xy2` for the second graph, `xy3` for the third, etc.)

As a bonus, that element already has the `pointer-events = all` so no need to set it.

Attaching to the click event, it does not seem to disrupt the ability to pan or zoom:

[Clicking on first graph.webm](https://user-images.githubusercontent.com/57320163/215351042-bf419867-7e60-4978-8efb-346691fbd299.webm)

Can you think of a scenario where this would interfere in the regular operation of the graph?  Is this too much of a hack or could it be added as configuration option for the card?  :pleading_face: 

Simply running this command does the trick to get the first graph to show the more-info when clicked:

```js
document
  .querySelector("home-assistant")
  .shadowRoot.querySelector("home-assistant-main")
  .shadowRoot.querySelector("app-drawer-layout")
  .querySelector("ha-panel-lovelace")
  .shadowRoot.querySelector("hui-root")
  .shadowRoot.querySelector("ha-app-layout")
  .querySelector("hui-masonry-view")
  .shadowRoot.querySelector("plotly-graph")
  .shadowRoot.querySelector("svg.main-svg g.xy:first-child").onclick =
  function (e) {
    console.log("Displaying more-info");
    const event = new Event("hass-more-info");
    event.detail = { entityId: "climate.rez_de_chaussee_chauffage_thermostat" };
    document.querySelector("home-assistant").dispatchEvent(event);
  };
```

**fixed formatting**

This is very close to what feature request #72 was :slightly_smiling_face:

**Reply by dbuezas** [reply]
Imo, this is too brittle to add to the official release, but by all means use it! (and make a Show&Tell post :) )

**Reply by dbuezas** [reply]
Click handlers added in official v3 release 🎉 
(see readme. I made good use of your snippet and ideas :) )

**Reply by dbuezas** [reply]
There are now hooks for `legend click` & `doubleclick`, `plot doubleclick` and trace click (which doesn't work and I don't know why).

**Reply by dbuezas** [reply]
And I just took a look at your code, VERY COOL! (also fixed the formatting so I could read it, hehe).

**Reply by MathieuIci** [reply]
Yay!  You're fast!  Congrats on the 3.0.0 release :partying_face: 

> Imo, this is too brittle to add to the official release, but by all means use it! (and make a Show&Tell post :) )

I agree that it's kind of a hack, attaching an event handler directly to the SVG created by plotly.  But for my use case it's exactly what I need and it works beautifully. (thanks for the formatting)

Unfortunately I don't think I can use it without modifying the card's source code.  I didn't dig, but it seems that after an entity's functions runs, the graph is recreated which deletes the event handler that was attached.

The yaml would need a configuration option like `function_to_run_after_graph_is_updated` lol.  That would be the best place to put in the code to attach the event handler.

**Reply by dbuezas** [reply]
I can add that, but please try first wrapping your code in a setTimeout(()=>{...},100) first.
And put the code after the entities so to ensure the parser is done fetching. E.g:
```yaml
entities:
  - entity: xxx
  - ...
run_last: $fn(.....
```

**Reply by octaviz** [reply]
> Awesome release, thank you!
> 
> I no longer need to use the lovelace-card-templater 🤓
> 
> I discovered this card a few days ago and have been having lots of fun, particularly with the "fn" filter:
> 
> ![image](https://user-images.githubusercontent.com/57320163/215238511-013fa246-b742-4911-ba8b-985913a0032a.png)
> 
> I have two questions:
> 
>     1. With this new beta release, I had to add `ha_theme: true` to my card config for the Home Assistant dark theme to be applied.  Is this expected?
> 
>     2. I've gone trough the documentation many times and searched online for an answer, but couldn't figure out if there are any way to define "tap" actions on title/annotations/icons (/anything) to have them perform more-info, service call, etc. ?  This would bring so many more possibilities, it's something I miss from the ApexCharts card.
> 
> 
> Edit: Hmm, I think something like [this feature enhancement request](https://github.com/dbuezas/lovelace-plotly-graph-card/issues/72) is what I am looking for.
> 
> Thanks again!

Hi!
How do you put the temperature value on the right side, thank you very much.
![imagen](https://user-images.githubusercontent.com/856579/219579722-96087c39-214f-4493-ba0f-5330c7a8390f.png)

**Reply by dbuezas** [reply]
Glad to hear! You should try `$ex` too then! (the same but shorter)
By the way, this is not in beta anymore, make sure to update to the latest stable release (v3.3.1 at the moment).
The value at the top is probably a nicely styled indicator. See [this](https://gist.github.com/dbuezas/92210c8fcaf8d86d6c2d358e9d83b52d#file-dashboard-yaml-example-yaml-L231) example

**Reply by dbuezas** [reply]
And `ha_theme: true` is set by default and only disabled if you use `raw_plotly_config: true`. I thought I had added that to the readme. I'll take a PR for the readme :)

**Reply by dbuezas** [reply]
For the standard click handlers see [this part of the documentation](https://github.com/dbuezas/lovelace-plotly-graph-card#entity-click-handlers), and for a completely over the top hacked-to-perfection solution see [this discussion](https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/219)

**Comment by dbuezas** [comment]
v3.0.0 (not beta) released!

**Reply by MathieuIci** [reply]
Mouhouhouhahaha, it works :smiley:!  I didn't have to use `setTimeout`, I simply put the function all the way at the bottom of the yaml.  You rock!

I created a bit of code that parses the list of configured entities and automatically adds event handlers without needing to manually specify each of them.

So now, by adding this at the end of the card yaml:

```
run_last: |-
  $fn ({getFromConfig}) => {
    const graphs = {};
    getFromConfig('entities').map(entity => {
      graphs[entity['entity']] = {
        graphElement: ('xaxis' in entity ? entity['xaxis'].replace(/^x1$/, 'x') : 'x') + ('yaxis' in entity ? entity['yaxis'].replace(/^y1$/, 'y') : 'y')
      };
    });
    const plotlyElement = document
      .querySelector("home-assistant").shadowRoot
      .querySelector("home-assistant-main").shadowRoot
      .querySelector("ha-panel-lovelace").shadowRoot
      .querySelector("hui-root").shadowRoot
      .querySelector("hui-masonry-view").shadowRoot
      .querySelector("plotly-graph").shadowRoot;
    Object.entries(graphs).forEach(([entity, config]) => {
      plotlyElement.querySelector(`svg.main-svg g.${config.graphElement} :first-child`).onclick = function (e) {
        const event = new Event("hass-more-info");
        event.detail = { entityId: entity };
        document.querySelector("home-assistant").dispatchEvent(event);
      };
    });
  }
```

Clicking on a graph pops up the more-info page corresponding to the entity using it.  Tested on multiple vertically stacked subplots but it should also work with horizontal stacks.

Only caveat for this "automatic event-handler-creation" to work is that a graph can only use a single entity, but can use many attributes from this entity.  It could be tweaked for other scenarios.

Your universal functions feature is so powerful :muscle:!

**Reply by dbuezas** [reply]
Very clever!
I didn't expect the `getFromConfig` param to be put to use so fast, I [use it internally](https://github.com/dbuezas/lovelace-plotly-graph-card/blob/master/src/parse-config/defaults.ts) for all the defaults that depend on what the user did or the fetched data (e.g unit of measurement) and during fetching. 
This is perfect for a Show and Tell!

Yeah, the universal functions took a lot of thought, but I think the result was worth it :)

**Comment by thecem** [comment]
I love this nice histogram funktions!!!

But in my case I´m getting this:

![grafik](https://user-images.githubusercontent.com/46648579/215578608-7716169f-ecc3-4546-aea9-001f7beba547.png)

Why is the yellow BG so smal and not alligned to the map?

I use your $fn tab example.

**Reply by dbuezas** [reply]
Let's move this to a new Q&A discussion. I'll respond there

**Reply by dbuezas** [reply]
Please post your yaml as text there too

**Reply by dbuezas** [reply]
And I agree, the histograms are really cool! :)
